src/esp/bindings_js/modules/replay_viewer.js

```diff
@@ -0,0 +1,373 @@
+    this.updateFpsFunction = setInterval(() => {
+      var duration;
+      var fps_desc;
+      if (window.performance.now) {
+        var timestamp = window.performance.now();
+        if (this.lastTimestamp) {
+          duration = (timestamp - this.lastTimestamp) / 1000.0;
+        }
+        this.lastTimestamp = timestamp;
+        fps_desc = "FPS";
+      } else {
+        duration = 0.5;
+        fps_desc = "APPROX FPS";
+      }
+
+      this.components.fps.innerHTML =
+        fps_desc +
+        " " +
+        (this.numRendersSinceLastUpdate / duration).toFixed(1).toString();
+      this.numRendersSinceLastUpdate = 0;
+    }, 500.0);
+
+    var filepath = "physics_demo.gfx_replay.json";
+    this.player = this.sim.sim
+      .getGfxReplayManager()
+      .readKeyframesFromFile(filepath);
+    if (!this.player) {
+      this.setErrorStatus("failed to load replay");
+    } else {
+      this.setStatus("");
+    }
+    if (this.player) {
+      this.player.setKeyframeIndex(0);
+    }
+  }
+
+  // PRIVATE methods.
+
+  setStatus(text) {
+    this.components.status.style = "color:white";
+    this.components.status.innerHTML = text;
+  }
+
+  setErrorStatus(text) {
+    this.components.status.style = "color:red";
+    this.components.status.innerHTML = text;
+  }
+
+  setWarningStatus(text) {
+    this.components.status.style = "color:orange";
+    this.components.status.innerHTML = text;
+  }
+
+  setSuccessStatus(text) {
+    this.components.status.style = "color:green";
+    this.components.status.innerHTML = text;
+  }
+
+  renderImage() {
+    this.sim.displayObservation("rgb");
+  }
+
+  render() {
+    this.renderImage();
+    this.numRendersSinceLastUpdate += 1;
+  }
+
+  removeActiveAction(action) {
+    const index = this.activeActions.indexOf(action);
+    if (index > -1) {
+      this.activeActions.splice(index, 1);
+    }
+  }
+
+  syncAgentRotation() {
+    let state = this.sim.getAgentState();
+
+    state.rotation = Module.quaternionMultiply(
+      Module.eulerToQuaternion([0, this.agentRotation, 0]),
+      Module.eulerToQuaternion([this.agentLookPitch, 0, 0])
+    );
+
+    this.sim.setAgentState(state);
+  }
+
+  initAgentState() {
+    let state = this.sim.getAgentState();
+    state.position = [-3.01, 0, 5.92];
+    this.sim.setAgentState(state);
+
+    this.agentRotation = 3.5;
+    this.agentLookPitch = -0.21;
+    this.syncAgentRotation();
+  }
+
+  handleAction(action) {
+    if (
+      action == "moveForward" ||
+      action == "moveBackward" ||
+      action == "strafeLeft" ||
+      action == "strafeRight"
+    ) {
+      let stepDist = 0.06;
+      let stepSignedDist;
+      let adjustedRotation;
+      if (action == "moveForward" || action == "moveBackward") {
+        stepSignedDist = action == "moveForward" ? stepDist : -stepDist;
+        adjustedRotation = -this.agentRotation - Math.PI * 0.5;
+      } else {
+        stepSignedDist = action == "strafeLeft" ? stepDist : -stepDist;
```

Comment:
Notice this design decision: because this app is not tied to the agent action API, I'm able to easily add new interactivity such as strafing without touching the C++.